src/ir/lowering/mod.rs

```diff
@@ -837,7 +848,10 @@ trait LowerClause {
 impl LowerClause for Clause {
     fn lower_clause(&self, env: &Env) -> Result<ir::ProgramClause> {
         let implication = env.in_binders(self.all_parameters(), |env| {
-            let consequence: ir::DomainGoal = self.consequence.lower(env)?;
+            let consequences: Vec<ir::DomainGoal> = self.consequence.lower(env)?;
+            assert_eq!(1, consequences.len());
+            let consequence = consequences.into_iter().next().unwrap();
```

Comment:
Ah I understand the problem. So this may not be the right way to do it, because at one point you might want to lower an "AST"-rule which would look like:
```
T: Iterator<Item = i32> :- blabla
```
but the problem is that the consequence part (left part) would be translated to: `ProjectionEq(<T as Iterator>::Item = i32), T: Iterator` so your assertion would fail.

Instead, we would need to output a vector of `ProgramClause` (again), and we would generate two clauses in my example:
```
ProjectionEq(<T as Iterator>::Item = i32) :- blabla
T: Iterator :- blabla
```

I think this is the right way, you might want to get @nikomatsakis's confirmation though.

Reply:
Good point. For `Clause` this is pretty easy.

For `Goal` it's.. a bit more involved. Bubbling up multiple leaf goals in `LowerGoal` would complicate the code quite a bit. I _think_ I could handle it with a `Goal::And` at the lowest level, but I'm not certain if that's sound.

Reply:
Meh. This "custom clause" stuff is mostly useful for writing unit tests of the SLG engine etc. I think it's fine to assert things here, though it'd be even nicer to move it to the grammar, so that a `Clause` always has a limited set of "domain goals" as its "consequence".